presto-main/src/main/java/com/facebook/presto/operator/scalar/MapZipWithFunction.java

```diff
@@ -0,0 +1,189 @@
+                ImmutableList.of(),
+                parseTypeSignature("map(K,V3)"),
+                ImmutableList.of(parseTypeSignature("map(K,V1)"), parseTypeSignature("map(K,V2)"), parseTypeSignature("function(K,V1,V2,V3)")),
+                false));
+    }
+
+    @Override
+    public boolean isHidden()
+    {
+        return false;
+    }
+
+    @Override
+    public boolean isDeterministic()
+    {
+        return false;
+    }
+
+    @Override
+    public String getDescription()
+    {
+        return "merge two maps into a single map by applying the lambda function to the pair of values with the same key";
+    }
+
+    @Override
+    public ScalarFunctionImplementation specialize(BoundVariables boundVariables, int arity, TypeManager typeManager, FunctionRegistry functionRegistry)
+    {
+        Type keyType = boundVariables.getTypeVariable("K");
+        Type inputValueType1 = boundVariables.getTypeVariable("V1");
+        Type inputValueType2 = boundVariables.getTypeVariable("V2");
+        Type outputValueType = boundVariables.getTypeVariable("V3");
+        Type outputMapType = typeManager.getParameterizedType(
+                StandardTypes.MAP,
+                ImmutableList.of(
+                    TypeSignatureParameter.of(keyType.getTypeSignature()),
+                    TypeSignatureParameter.of(outputValueType.getTypeSignature())));
+        return new ScalarFunctionImplementation(
+                false,
+                ImmutableList.of(
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
+                        valueTypeArgumentProperty(RETURN_NULL_ON_NULL),
+                        functionTypeArgumentProperty(MapZipWithLambda.class)),
+                METHOD_HANDLE.bindTo(keyType).bindTo(inputValueType1).bindTo(inputValueType2).bindTo(outputMapType),
+                Optional.of(STATE_FACTORY.bindTo(outputMapType)),
+                isDeterministic());
+    }
+
+    public static Object createState(MapType mapType)
+    {
+        return new PageBuilder(ImmutableList.of(mapType));
+    }
+
+    public static Block mapZipWith(
+            Type keyType,
+            Type inputValueType1,
+            Type inputValueType2,
+            MapType outputMapType,
+            Object state,
+            Block block1,
+            Block block2,
+            MapZipWithLambda function)
+    {
+        SingleMapBlock mapBlock1 = (SingleMapBlock) block1;
+        SingleMapBlock mapBlock2 = (SingleMapBlock) block2;
+        Type outputValueType = outputMapType.getValueType();
+
+        PageBuilder pageBuilder = (PageBuilder) state;
+        if (pageBuilder.isFull()) {
+            pageBuilder.reset();
+        }
+        BlockBuilder mapBlockBuilder = pageBuilder.getBlockBuilder(0);
+        BlockBuilder blockBuilder = mapBlockBuilder.beginBlockEntry();
+
+        boolean[] keyFound = new boolean[mapBlock2.getPositionCount()];
```

Comment:
If `seekKey` has `O(1)` complexity, is it possible to avoid using this array?

Reply:
Add a comment to explain why this is likely worse having.